Add a document about setting up `nerdctl build` with BuildKit (#928)
containerd/nerdctl: docs/build.md

# Setting up `nerdctl build` with BuildKit

`nerdctl build` (and `nerdctl compose build`) relies on [BuildKit](https://github.com/moby/buildkit).
To use it, you need to set up BuildKit.

BuildKit has 2 types of backends.

- **containerd worker**: BuildKit relies on containerd to manage containers and images, etc. containerd needs to be up-and-running on the host.
- **OCI worker**: BuildKit manages containers and images, etc. containerd isn't needed. This worker relies on runc for container execution.

You need to set up BuildKit with either of the above workers.

> Note that OCI worker cannot access images managed by containerd.

Reviewer: ```suggestion
Note that OCI worker cannot access base images (`FROM` images in Dockerfiles)  managed by containerd.
```

Thus you cannot let `nerdctl build` use containerd-managed images as the base image.
They includes:

> - images pulled using `nerdctl pull`

Reviewer: This is not a problem, typically

- images previously built with `nerdctl build`

For example, the following build fails with OCI worker because it tries to use containerd-managed image `mybase`.

```console
$ nerdctl pull ghcr.io/stargz-containers/ubuntu:20.04-org
$ nerdctl tag ghcr.io/stargz-containers/ubuntu:20.04-org mybase
$ mkdir -p /tmp/ctx && cat <<EOF > /tmp/ctx/Dockerfile
FROM mybase
RUN echo hello
EOF
$ nerdctl build -t foo /tmp/ctx
```

This limitation can be avoided using containerd worker as mentioned later.

## Setting up BuildKit with containerd worker

### Rootless (requires BuildKit v0.10.0 or later)

```
$ CONTAINERD_NAMESPACE=default containerd-rootless-setuptool.sh install-buildkit-containerd
```

> `containerd-rootless-setuptool.sh` is aware of `CONTAINERD_NAMESPACE` and `CONTAINERD_SNPASHOTTER` envvars.

Reviewer: ```suggestion
`containerd-rootless-setuptool.sh` is aware of `CONTAINERD_NAMESPACE` and `CONTAINERD_SNAPSHOTTER` envvars.
```

It installs buildkitd to the specified containerd namespace.
This allows BuildKit using containerd-managed images in that namespace as the base image.
Note that BuildKit can't use images in other namespaces as of now.

> If `CONTAINERD_NAMESPACE` envvar is not specified, this script installs buildkitd to "buildkit" namespace (not "default" namespace).

Reviewer: ```suggestion
If `CONTAINERD_NAMESPACE` envvar is not specified, this script configures buildkitd to use "buildkit" namespace (not "default" namespace).
```
